orttraining/orttraining/core/optimizer/gather_replacement.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+// gather_replacement.cc
+
+#include "orttraining/core/optimizer/gather_replacement.h"
+
+#include "core/common/logging/logging.h"
+#include "core/optimizer/rewrite_rule.h"
+#include "core/optimizer/utils.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+namespace {
+
+const char* NumSegmentsNodeArgName = "num_segments";
+const char* SegmentOffsetssNodeArgName = "segment_offsets";
+const char* LastSegmentPartialSegmentCountNodeArgName = "last_segment_partial_segment_count";
+const char* LastSegmentPartialSegmentOffsetNodeArgName = "last_segment_partial_segment_offset";
+const char* PerSegmentPartialSegmentCountsNodeArgName = "per_segment_partial_segment_counts";
+const char* PerSegmentPartialSegmentOffsetsNodeArgName = "per_segment_partial_segment_offsets";
+const char* DXIndeicesSortedNorArgName = "dX_indices_sorted";
+const char* DYIndeicesSortedNorArgName = "dY_indices_sorted";
+const char* GatherInternalName("GatherInternal");
```

Comment:
nit:
"SegmentOffsetssNodeArgName" -> "SegmentOffsetsNodeArgName"
"DXIndeicesSortedNorArgName" -> "DXIndicesSortedNodeArgName"
"DYIndeicesSortedNorArgName" -> "DYIndicesSortedNodeArgName"
inconsistent initialization of GatherInternalName